### ISSUE TITLE
[Question] How to set the priors of parameters for GP?

### ISSUE THREAD

**Issue by songlei00** [pytorch/botorch#1837]
Hello, I am using `SingleTaskGP` to fit a cos function. I find that the priors of parameters have a great influence on the performance. 

I write a demo. The blue dashed line is the objective. If I use the **default configs in botorch**, I will get a perfect fitting (the `model_1` in the following code and the green line in the figure). If I use the **default configs in gpytorch**, the fitting is very poor (the `model_2` in the following code and the red line in the figure). The code example and running result are as follows.

## Code example

``` python
import torch
import gpytorch
from gpytorch.kernels import ScaleKernel, MaternKernel
from botorch import fit_gpytorch_mll
from botorch.models import SingleTaskGP
import matplotlib.pyplot as plt

torch.manual_seed(0)

def func(x):
    true_f = torch.cos(x * 15) - torch.abs(x)
    return true_f.reshape(-1)

def generate(n):
    x = torch.rand(n, 1).double() 
    y = func(x)
    return x, y

# generate data
train_x, train_y = generate(10)
plot_x = torch.linspace(0, 1, 50).double()
plot_y = func(plot_x).reshape(-1)

# model 1
model_1 = SingleTaskGP(train_x, train_y.reshape(-1, 1))
mll_1 = gpytorch.mlls.ExactMarginalLogLikelihood(model_1.likelihood, model_1)
fit_gpytorch_mll(mll_1)

# model 2
covar_module = ScaleKernel(MaternKernel())
model_2 = SingleTaskGP(train_x, train_y.reshape(-1, 1), covar_module=covar_module)
mll_2 = gpytorch.mlls.ExactMarginalLogLikelihood(model_2.likelihood, model_2)
fit_gpytorch_mll(mll_2)

with torch.no_grad():
    _, ax = plt.subplots()
    ax.plot(plot_x, plot_y, color='blue', linestyle='--')

    # model 1
    pred_dist_1 = model_1.likelihood(model_1(plot_x))
    y_1 = pred_dist_1.mean
    std_1 = pred_dist_1.variance.sqrt()
    ax.plot(plot_x, y_1, color='green')
    ax.fill_between(plot_x.reshape(-1), y_1 - std_1, y_1 + std_1, color='green', alpha=0.5)
    # model 2
    pred_dist_2 = model_2.likelihood(model_2(plot_x))
    y_2 = pred_dist_2.mean
    std_2 = pred_dist_2.variance.sqrt()
    ax.plot(plot_x, y_2, color='red')
    ax.fill_between(plot_x.reshape(-1), y_2 - std_2, y_2 + std_2, color='red', alpha=0.5)

    plt.savefig('test.png')
```

## Result

![image](https://github.com/pytorch/botorch/assets/44307424/be534846-abdd-45d3-a31a-b77f4fad2c32)

## Questions

It is very confusing, so I dive into the source code and I find the `SingleTaskGP` use the following code to create the default kernel for GP (the following code is from the line 150-159 in botorch.models.gp_regression.py). It sets magic priors for `MaternKernel` and `ScaleKernel`, which is `GammaPrior(3.0, 6.0)` and `GammaPrior(2.0, 0.15)` respectively. 

I want to know why you select such priors with magic numbers 3.0, 6.0, 2.0 and 0.15? Is there a practial guidance to tune the hyper-parameters?

``` python
covar_module = ScaleKernel(
    MaternKernel(
        nu=2.5,
        ard_num_dims=transformed_X.shape[-1],
        batch_shape=self._aug_batch_shape,
        lengthscale_prior=GammaPrior(3.0, 6.0),
    ),
    batch_shape=self._aug_batch_shape,
    outputscale_prior=GammaPrior(2.0, 0.15),
)
```

## System Info
Please provide information about your setup, including
- BoTorch Version 0.8.2
- GPyTorch Version 1.9.1
- PyTorch Version 1.12.0
- Ubuntu 18.04

**Comment by Balandat** [comment]
Hi, thanks for your question. As you observed, the priors on the GP parameters can have a large effect on the model performance, especially in the low data regime (with more data the effect of the priors becomes less important).

This is one of the dirty secrets of GPs. There are a few things you can do to choose reasonable priors - e.g. in BoTorch we typically assume that the inputs have been normalized (to the unit cube) and the observations have been standardized (zero mean, unit variance), which gives us reasonable ranges for what we'd expect lengthscale and outputscale of the parameters to be, so we can choose a prior that doesn't have much support outside of that. 

Beyond that, the way you'd choose the priors depends a bit on how general you want things to be. If you work in a particular domain and have a lot of data to work with, you could do this in an empirical fashion: You would fit models with uninformative priors on larger amounts of data, and then observe what the distribution of the parameters across a large number of examples are. You can then fit some functional form priors to those empirically observed parameter distributions and use those in the model. 

We did something similar for coming up with our default priors. We actually think that the priors on the noise are off (they put too much weight on highly noisy scenarios) and we're planning to change those down the road to get better performance in less noisy settings.
